src/storage/conversion/ChainedConverter.ts

```diff
@@ -1,89 +1,305 @@
 import type { Representation } from '../../ldp/representation/Representation';
+import type { ValuePreferences } from '../../ldp/representation/RepresentationPreferences';
 import { getLoggerFor } from '../../logging/LogUtil';
-import { InternalServerError } from '../../util/errors/InternalServerError';
-import { matchesMediaType } from './ConversionUtil';
+import { BadRequestHttpError } from '../../util/errors/BadRequestHttpError';
+import { NotImplementedHttpError } from '../../util/errors/NotImplementedHttpError';
+import { hasMatchingMediaTypes, matchesMediaType } from './ConversionUtil';
+import { RepresentationConverter } from './RepresentationConverter';
 import type { RepresentationConverterArgs } from './RepresentationConverter';
-import { TypedRepresentationConverter } from './TypedRepresentationConverter';
+import type { TypedRepresentationConverter } from './TypedRepresentationConverter';
+
+type ConverterPreference = {
+  converter: TypedRepresentationConverter;
+  type: string;
+  weight: number;
+};
+
+type ConversionPath = {
+  converters: TypedRepresentationConverter[];
+  intermediateTypes: string[];
+  inTypes: ValuePreferences;
+  outTypes: ValuePreferences;
+};
+
+type MatchedPath = {
+  path: ConversionPath;
+  inType: string;
+  outType: string;
+  weight: number;
+};
+
+/**
+ * Finds the best media type that matches both the given `ValuePreferences`.
+ * It is assumed that `left` contains the more specific types than `right` if there are differences,
+ * since `left` usually corresponds to a converters output and `right` to the next converters input.
+ * "Best" implies the highest weight, which will also be returned.
+ */
+function getMatchingType(left: ValuePreferences, right: ValuePreferences): { type: string; weight: number } {
+  let bestMatch: { type: string; weight: number } = { type: 'invalid', weight: 0 };
+
+  // Try to find the matching type with the best weight
+  const rightEntries = Object.entries(right);
+  for (const [ leftType, leftWeight ] of Object.entries(left)) {
+    if (leftWeight <= bestMatch.weight) {
+      continue;
+    }
+    for (const [ rightType, rightWeight ] of rightEntries) {
+      const weight = leftWeight * rightWeight;
+      if (weight > bestMatch.weight && matchesMediaType(leftType, rightType)) {
+        bestMatch = { type: leftType, weight };
+        if (weight === 1) {
```

Comment:
We are comparing floating numbers here, so I think

```suggestion
        if (weight >= 1) {
```

or perhaps even

```suggestion
        if (weight >= 0.9999) {
```

Reply:
This is an optimization to jump out of the loop when we find the maximum possible value (which is 1), so it's ok if this is never reached due to float multiplication. Multiplying floats which all have value 1 should still result in 1 afaik, which is the only relevant case, unless I'm misremembering.